# Tasks

https://github.com/ianmcloughlin/computational_theory/blob/main/assessment/tasks.md

## Task 1: Binary Representations

This task involves implementing four binary manipulation functions commonly used in cryptographic algorithms:

- Rotate Left (rotl)

- Rotate Right (rotr)

- Choose (ch)

- Majority (maj)

In [73]:
hexNum = 0x1010002
# Test value that will be used against functions

### Functions

### Rotate Left (rotl)

The rotl function rotates the bits in a 32-bit unsigned integer to the left by n places. Bits that "fall off" the left end reappear at the right end.

The function takes two parameters:
- `x`: The 32-bit unsigned integer to be rotated
- `n`: The number of bit positions to rotate left (default is 1)

For example, when rotating 0x1010002 left by 4 bits:
1. The 4 leftmost bits move to the rightmost positions
2. All other bits shift left by 4 positions
3. The result is masked with 0xFFFFFFFF to ensure it remains a 32-bit value

This operation is commonly used in cryptographic algorithms like SHA-256, where bit manipulation helps create complex transformations of the input data.

In [74]:
def rotl(x, n=1):
    # Rotate left: Rotates the bits in a 32-bit integer to the left n places.
    # Ensure x is a 32-bit integer
    x = x & 0xFFFFFFFF
    
    # Perform rotation
    return ((x << n) | (x >> (32 - n))) & 0xFFFFFFFF

print(f"hexNum: {hexNum:08x}")
print(f"Rotated left 4 bits: {rotl(hexNum, 4):08x}")
print(f"Rotated left 8 bits: {rotl(hexNum, 8):08x}")

hexNum: 01010002
Rotated left 4 bits: 10100020
Rotated left 8 bits: 01000201


### Rotate Right (rotr)

The rotr function rotates the bits in a 32-bit unsigned integer to the right by n places. Bits that "fall off" the right end reappear at the left end.

The function takes two parameters:
- `x`: The 32-bit unsigned integer to be rotated
- `n`: The number of bit positions to rotate right (default is 1)

This is the opposite of the rotl function and is also commonly used in cryptographic algorithms like SHA-256.

In [75]:
def rotr(x, n=1):
    # Rotate right: Rotates the bits in a 32-bit integer to the right n places.
    # Ensure x is a 32-bit integer
    x = x & 0xFFFFFFFF
    
    # Perform rotation
    return ((x >> n) | (x << (32 - n))) & 0xFFFFFFFF

print(f"hexNum: {hexNum:08x}")
print(f"Rotated right 4 bits: {rotr(hexNum, 4):08x}")
print(f"Rotated right 8 bits: {rotr(hexNum, 8):08x}")

hexNum: 01010002
Rotated right 4 bits: 20101000
Rotated right 8 bits: 02010100


### Choose (ch)

The ch function selects bits from y where x has bits set to 1, and bits from z where x has bits set to 0.

For each bit position i:
- If bit i in x is 1, the output bit i comes from y
- If bit i in x is 0, the output bit i comes from z

This function is commonly used in cryptographic hash functions like SHA-256. It acts as a bit-by-bit multiplexer, using each bit of x to choose between the corresponding bits of y and z.

The implementation uses the following operations:
- (x & y): Keep bits from y where x has 1s
- (~x & z): Keep bits from z where x has 0s
- XOR (^): Combine both parts into the final result

When applied to a 32-bit input, this creates complex bit interdependencies that contribute to the avalanche effect essential for secure hash functions.

In [76]:
def ch(x, y, z):
    # For each bit position i, if x has bit i set, then output bit i from y, 
    # else output bit i from z.
    return (x & y) ^ (~x & z)

x1, y1, z1 = hexNum, 0xAAAAAAAA, 0x55555555
print(f"x: {x1:08x}, y: {y1:08x}, z: {z1:08x}")
print(f"ch(x, y, z): {ch(x1, y1, z1):08x}")

x: 01010002, y: aaaaaaaa, z: 55555555
ch(x, y, z): 54545557


### Majority (maj)

The maj function takes a majority vote of bits in x, y, and z. The output has a 1 in bit position i where at least two of the three inputs have 1's in position i.

The function takes three parameters:
- `x`, `y`, and `z`: The three 32-bit unsigned integers to vote on

The implementation calculates this using the formula:
- `(x & y) ^ (x & z) ^ (y & z)`

This efficiently determines, for each bit position:
- If both x and y have a 1, OR
- If both x and z have a 1, OR
- If both y and z have a 1

This function is a critical component in the SHA-256 hash algorithm, where it helps create complex dependencies between bits during the message compression phase.

In [77]:
def maj(x, y, z):
    # For each bit position i, output a 1 if at least two of the 
    # three inputs have a 1 in position i, and 0 otherwise.
    return (x & y) ^ (x & z) ^ (y & z)

x1, y1, z1 = hexNum, hexNum, 0x00000000
print(f"x: {x1:08x}, y: {y1:08x}, z: {z1:08x}")
print(f"maj(x, y, z): {maj(x1, y1, z1):08x}")

x: 01010002, y: 01010002, z: 00000000
maj(x, y, z): 01010002


## Task 2: Hash Functions

### Introduction
A hash function is a mathematical algorithm that maps data of arbitrary size to a fixed-size value. Hash functions are widely used in computer science for tasks such as:

- Data retrieval in hash tables
- Cryptography
- Data integrity checks

The output of a hash function is called a hash value or hash code. A good hash function should:

- Minimize collisions (different inputs producing the same hash value)
- Be efficient to compute
- Distribute hash values uniformly across the output range

### The K&R Hash Function
The following hash function is taken from "The C Programming Language" by Brian Kernighan and Dennis Ritchie. It is a simple and efficient hash function designed for strings.

#### Algorithm Explanation:
1. **Initialization**: Start with hashval = 0
2. **Iterate through the string**: For each character in the string:
   - Multiply the current hashval by 31
   - Add the ASCII value of the current character
3. **Modulo operation**: After processing the string, take the modulo of the hash value with 101 to ensure the result fits within a specific range

#### C Programming Language
```c
// Computes a hash value for a string
unsigned hash(char *s) {
    unsigned hashval;
    for (hashval = 0; *s != '\0'; s++)
        hashval = *s + 31 * hashval;
    return hashval % 101;
}

### Source

Kernighan, B. W., & Ritchie, D. M. (1988). The C Programming Language (2nd ed.), Section 6.3, “Hashing”

- https://www.cs.sfu.ca/~ashriram/Courses/CS295/assets/books/C_Book_2nd.pdf
- https://search.ebscohost.com/login.aspx?direct=true&AuthType=ip,sso&db=cat10672a&AN=atu.KOHA.ATU.408073316&site=eds-live&scope=site&custid=s2873033

#### Python Version

In [78]:
def hash_function(s):
    hashval = 0
    for c in s:
        hashval = ord(c) + 31 * hashval
    return hashval % 101

# Test function with example strings
exampleWords = ["hello", "world", "python", "programming", "hash", "function", "9"]
for s in exampleWords:
    print(f"Hash of '{s}': {hash_function(s)}")

Hash of 'hello': 17
Hash of 'world': 34
Hash of 'python': 91
Hash of 'programming': 89
Hash of 'hash': 15
Hash of 'function': 100
Hash of '9': 57


### Why values 31 and 101?

#### The Value 31
- **Prime Number**: 31 is a prime number, which helps distribute hash values more evenly
- **Near Power of 2**: 31 is close to 32 (2^5), which makes multiplication efficient on computers (31 * n = 32 * n - n = n << 5 - n)
- **Empirical Performance**: In practice, 31 has been found to produce fewer collisions for typical string data
- **Balance**: It's large enough to create good distribution but small enough to avoid overflow in 32-bit systems

#### The Value 101
- **Prime Modulus**: 101 is a prime number, which is ideal for the modulo operation in hash functions
- **Table Size**: This suggests the original implementation was designed for a hash table with 101 slots
- **Collision Reduction**: Prime moduli help distribute hash values more evenly across the hash table, reducing collisions
- **Appropriate Scale**: 101 is large enough to be useful but small enough to be practical in the original C implementation

The combination of multiplying by a prime (31) and taking the modulus by another prime (101) helps create a good distribution of hash values, which is essential for hash table performance.

## Task 3: SHA256

### SHA-256 Description

*Some of the next tasks will involve SHA-256, so here’s a brief description about it:*

SHA-256 (Secure Hash Algorithm 256) is a cryptographic hash function in the SHA-2 family, defined by NIST in FIPS 180-4. It processes input data of any length and produces a 256-bit (32-byte) fixed-size output, typically rendered as a 64-character hexadecimal string. SHA-256 offers strong collision resistance, preimage resistance, and second-preimage resistance, making it a cornerstone of modern security protocols—used in everything from SSL/TLS to blockchain consensus mechanisms.

### Key Features

1. **Fixed Output Size**:  
    Always produces a 256-bit hash value.

2. **Deterministic**:  
    Same input always produces the same hash.

3. **Collision Resistance**:  
    Infeasible to find two inputs with the same hash.

4. **Preimage Resistance**:  
    Infeasible to reverse-engineer the input from the hash.

5. **Second-Preimage Resistance**:  
    Infeasible to find a different input with the same hash.

6. **Avalanche Effect**:  
    Small input changes result in drastically different hashes.

7. **Wide Adoption**:  
    Used in SSL/TLS, digital signatures, blockchain, and more.

### Applications

- **Data Integrity Verification**:  
  Ensures data has not been altered during transmission or storage.

- **Digital Signatures**:  
  Verifies authenticity and integrity of digital documents.

- **Password Hashing**:  
  Protects stored passwords by hashing them.

- **Blockchain**:  
  Used in mining and transaction verification.

- **File Integrity Checks**:  
  Verifies file integrity during downloads or transfers.

### Source

- https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf
- https://csrc.nist.gov/glossary/term/sha_2

### Introduction
SHA256 (Secure Hash Algorithm 256-bit) is a cryptographic hash function that produces a 256-bit (32-byte) hash value. Before computing the hash, a specific padding scheme is applied to ensure the message length is a multiple of 512 bits.

### Padding Requirements
The SHA256 padding procedure consists of three components:

1. **A single '1' bit**: Appended immediately after the message
2. **Zero bits**: Added until the message length is congruent to 448 modulo 512 bits
3. **64-bit length field**: The original message length in bits as a 64-bit big-endian integer

This ensures the total length is a multiple of 512 bits, which is required for the SHA256 algorithm to process the message in fixed-size blocks.

### Implementation
The following functions calculate and display the padding that would be applied to a file:

In [79]:
def calculate_sha256_padding(file_path):
    # Get file size in bits
    with open(file_path, 'rb') as f:
        # Get file size in bytes
        f.seek(0, 2)  # Go to the end of the file
        file_size_bytes = f.tell()
    
    # Convert bytes to bits
    file_size_bits = file_size_bytes * 8
    
    # Initialize padding with a single '1' bit followed by 7 zero bits
    padding = bytearray([0x80])
    
    # Calculate how many zero bytes to add
    # Message length needs to be congruent to 448 modulo 512 bits
    # That's 56 bytes modulo 64 bytes
    zero_bytes = (56 - ((file_size_bytes + 1) % 64)) % 64
    
    # Add zero bytes
    padding.extend([0] * zero_bytes)
    
    # Add the original message length as a 64-bit integer
    for i in range(8):
        padding.append((file_size_bits >> (56 - i * 8)) & 0xFF)
    
    return padding

In [80]:
def display_sha256_padding(padding):
    hexValues = [f"{b:02x}" for b in padding]
    
    # Format with spaces and newlines for readability
    formatted = ""
    for i, hexVal in enumerate(hexValues):
        formatted += hexVal + " "
        if (i + 1) % 26 == 0:  # 26 bytes per line for readability
            formatted += "\n"
    
    print(formatted.strip())

In [81]:
def sha256_padding(file_path):
    padding = calculate_sha256_padding(file_path)
    display_sha256_padding(padding)

### Explanation of the SHA256 Padding Process

The SHA256 padding ensures the message length is a multiple of 512 bits by adding:

1. **A single '1' bit**: This is represented as the byte 0x80 (binary 10000000)
2. **Zero bits**: Added until the message length is congruent to 448 modulo 512 bits
3. **64-bit length field**: The original message length in bits encoded as a big-endian integer

#### Example Calculation

For a file containing "abc" (3 bytes = 24 bits):

- First byte of padding: 0x80 (the '1' bit followed by 7 zero bits)
- Zero bytes needed: 52 bytes
  - Original message = 3 bytes
  - We've added 1 byte (the 0x80)
  - Target is 56 bytes (448 bits) before the length field
  - Calculation: (56 - ((3 + 1) % 64)) % 64 = 52 zero bytes
- Length field: 24 bits = 0x0000000000000018 (8 bytes)

The complete padding (61 bytes total):

In [82]:
# Create a test file containing 'abc' (tasks.md example specification) 
# and calculate its SHA256 padding
with open('resources/test_file.txt', 'w') as f:
    f.write('abc')

sha256_padding('resources/test_file.txt')

80 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 18


## Task 4: Prime Numbers

### Introduction
A prime number is a natural number greater than 1 that cannot be formed by multiplying two smaller natural numbers. Prime numbers are fundamental to various fields including cryptography, where they serve as the building blocks for encryption algorithms like RSA.

This task implements two different methods to find prime numbers:
1. **Trial division** - A straightforward approach checking divisibility.
2. **Precomputed list** - A lookup-based approach for efficiency.

Both algorithms have different performance characteristics and use cases in computational applications.

### Why We Like Primes

Prime numbers are fundamental to number theory because of the Fundamental Theorem of Arithmetic, which states that:

> Every integer greater than 1 can be represented uniquely as a product of prime numbers (up to the order of the factors).

For example:
- 12 = 2² × 3
- 60 = 2² × 3 × 5
- 2310 = 2 × 3 × 5 × 7 × 11

This theorem is why prime numbers are considered the "building blocks" of all natural numbers. This property makes them essential for:

- **Cryptography**: RSA encryption relies on the difficulty of factoring large numbers into their prime components.
- **Hash functions**: Prime numbers help create well-distributed hash values.
- **Random number generation**: Prime-based algorithms produce high-quality pseudo-random sequences.
- **Error correction codes**: Prime-based codes are used for reliable data transmission.

### Source
- https://en.wikipedia.org/wiki/Fundamental_theorem_of_arithmetic

In [83]:
# Method 1: Check if a number is prime by iterating through possible divisors
print("Method 1")

# List of numbers to check
numbers = [1, 2, 5, 10, 343]

for num in numbers:
    # Negative numbers, 0 and 1 are not primes
    if num > 1:
        # Iterate from 2 to n // 2
        for i in range(2, (num // 2) + 1):
            # If num is divisible by any number between 2 and n / 2, it is not prime
            if (num % i) == 0:
                print(num, "is not a prime number")
                break
        else:
            print(num, "is a prime number")
    else:
        print(num, "is not a prime number")

Method 1
1 is not a prime number
2 is a prime number
5 is a prime number
10 is not a prime number
343 is not a prime number


### Explanation of Method 1

The first approach uses trial division to check if a number is prime.

A number is prime if it's greater than 1 and cannot be divided evenly by any number from 2 to n/2.

The algorithm checks each potential divisor and breaks as soon as one is found.

Time complexity: O(n) for each number being tested.

This method works well for small numbers but becomes inefficient for larger values.

### Source

- https://www.geeksforgeeks.org/python-program-to-check-whether-a-number-is-prime-or-not/

In [84]:
print("\nMethod 2")

# Method 2: Check if 'num' is a prime number by checking its presence in 'primeNos'
primeNos = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997]

# Display only the first 10 primes and their status
for num in primeNos[:15]:
    is_prime = num in primeNos
    print(f"Is {num} a prime number? {is_prime}")


Method 2
Is 2 a prime number? True
Is 3 a prime number? True
Is 5 a prime number? True
Is 7 a prime number? True
Is 11 a prime number? True
Is 13 a prime number? True
Is 17 a prime number? True
Is 19 a prime number? True
Is 23 a prime number? True
Is 29 a prime number? True
Is 31 a prime number? True
Is 37 a prime number? True
Is 41 a prime number? True
Is 43 a prime number? True
Is 47 a prime number? True


### Explanation of Method 2

The second approach uses a pre-computed list of prime numbers.

This method is extremely fast (O(1) lookup time) but only works for numbers within the range of the list.

The list includes the first 168 prime numbers up to 997.

This approach is commonly used in cryptographic applications where known prime numbers are needed.

## Task 5: Roots

In [85]:
import math

def first_32_bits_of_fractional_part(number):
    # Extract the fractional part of the number
    fractional_part = number - math.floor(number)
    # Convert the fractional part to the first 32 bits
    first_32_bits = int(fractional_part * (2**32))
    return first_32_bits

In [86]:
def is_prime(n):
    # Check if a number is prime
    if n <= 1:
        return False
    if n <= 3:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    return True

In [87]:
def first_n_primes(n):
    # Generate the first n prime numbers
    primes = []
    num = 2
    while len(primes) < n:
        if is_prime(num):
            primes.append(num)
        num += 1
    return primes

In [88]:
def taskFive():
    # Calculate the first 32 bits of the fractional part of the square root of the first 100 primes
    primes = first_n_primes(100)
    results = {}
    for prime in primes:
        sqrt_fractional_bits = first_32_bits_of_fractional_part(math.sqrt(prime))
        results[prime] = f"0x{sqrt_fractional_bits:08x}"  # Format as hexadecimal
    return results

In [89]:
# Display the results
results = taskFive()
for prime, hex_bits in results.items():
    print(f"Prime: {prime}, First 32 bits of fractional part of sqrt: {hex_bits}")

Prime: 2, First 32 bits of fractional part of sqrt: 0x6a09e667
Prime: 3, First 32 bits of fractional part of sqrt: 0xbb67ae85
Prime: 5, First 32 bits of fractional part of sqrt: 0x3c6ef372
Prime: 7, First 32 bits of fractional part of sqrt: 0xa54ff53a
Prime: 11, First 32 bits of fractional part of sqrt: 0x510e527f
Prime: 13, First 32 bits of fractional part of sqrt: 0x9b05688c
Prime: 17, First 32 bits of fractional part of sqrt: 0x1f83d9ab
Prime: 19, First 32 bits of fractional part of sqrt: 0x5be0cd19
Prime: 23, First 32 bits of fractional part of sqrt: 0xcbbb9d5d
Prime: 29, First 32 bits of fractional part of sqrt: 0x629a292a
Prime: 31, First 32 bits of fractional part of sqrt: 0x9159015a
Prime: 37, First 32 bits of fractional part of sqrt: 0x152fecd8
Prime: 41, First 32 bits of fractional part of sqrt: 0x67332667
Prime: 43, First 32 bits of fractional part of sqrt: 0x8eb44a87
Prime: 47, First 32 bits of fractional part of sqrt: 0xdb0c2e0d
Prime: 53, First 32 bits of fractional part 

### Explanation of Root

We identify prime numbers using a prime-checking function.

For each prime number, we calculate its square root using Python's ```math.sqrt()``` function.

We extract just the fractional part by subtracting the integer portion.

This fractional part is then converted to the first 32 bits by:
- Multiplying by 2^32
- Converting to an integer (truncating any remaining fraction)

### Why This Matters:

- These constants are used in SHA-256 and other cryptographic hash functions.
- Using the fractional parts of square roots of primes provides "nothing up my sleeve" numbers.
- These values are believed to have no hidden structure that could weaken the hash function.
- The first 8 values correspond to the initial hash values (H0 to H7) in the SHA-256 algorithm.

The output is displayed in hexadecimal for readability and to match how these constants are typically represented in hash function specifications.

## Task 6: Proof of Work

In this task, we aim to identify English words whose SHA-256 hash digests begin with the greatest number of zero bits. This process is analogous to a simplified proof-of-work system, where finding a digest with a certain number of leading zeros demonstrates computational effort.

In [90]:
import hashlib
import urllib.request
import random

In [91]:
def get_english_words(url="https://raw.githubusercontent.com/dwyl/english-words/master/words_alpha.txt"):
    # Get words from a GitHub dictionary repository
    with urllib.request.urlopen(url) as response:
        content = response.read().decode('utf-8')
    
    # Split into words and filter out empty strings
    words = [word.strip() for word in content.splitlines() if word.strip()]
    return words

- We fetch a large open-source English dictionary (words_alpha.txt).
    - Originally used a local file called `words.txt` to do this task, changed to using a dictionary from a GitHub repository as it made my repo look cleaner.

- Each line is a word we normalize by stripping whitespace.

### Source

- https://github.com/dwyl/english-words

In [92]:
def count_leading_zero_bits(hash_bytes):
    # Count leading zero bits in a hash byte array
    zero_bits = 0
    
    # Process each byte
    for byte in hash_bytes:
        if byte == 0:
            zero_bits += 8
            continue
            
        # Count leading zeros in non-zero byte
        mask = 0x80  # 10000000 in binary
        while byte & mask == 0:
            zero_bits += 1
            mask >>= 1
            
        break  # Once we find a 1 bit, we're done
        
    return zero_bits

- hash_hex is the 64-character hexadecimal SHA-256 digest.

- We convert it to a 256-length binary string and count leading 0 bits.

In [93]:
def words_with_most_leading_zeros(word_list):
    # Find words with the most leading zero bits in their SHA256 hash
    max_zeros = 0
    best_words = []
    
    for word in word_list:
        word = word.strip().lower()
        hash_hex = hashlib.sha256(word.encode()).hexdigest()
        # Convert hex string to byte array
        hash_bytes = bytes.fromhex(hash_hex)
        leading_zeros = count_leading_zero_bits(hash_bytes)
        
        if leading_zeros > max_zeros:
            max_zeros = leading_zeros
            best_words = [word]
        elif leading_zeros == max_zeros and leading_zeros > 0:
            best_words.append(word)
    
    return best_words, max_zeros

- We normalize each word (lowercase, no extra spaces).

- Compute its SHA-256 digest and count leading zeros.

- Track the maximum zero count and corresponding words.

In [94]:
def taskSix():
    # Get random words
    all_words = get_english_words()
    print(f"Downloaded {len(all_words)} English words")
    
    # If the dictionary is too large, use a random sample
    sample_size = min(50000, len(all_words))
    print(f"Using {sample_size} words")
    word_sample = random.sample(all_words, sample_size)
    
    # Find words with most leading zeros
    best_words, max_zeros = words_with_most_leading_zeros(word_sample)
    
    # Display results
    print(f"\nWords with most leading zero bits ({max_zeros} bits):")
    for word in best_words:
        hash_hex = hashlib.sha256(word.encode()).hexdigest()
        print(f"Word: '{word}'")
        print(f"SHA256: {hash_hex}")
        print(f"Binary prefix: {bin(int(hash_hex[:16], 16))[2:].zfill(64)[:32]}...")

taskSix()

Downloaded 370105 English words
Using 50000 words

Words with most leading zero bits (16 bits):
Word: 'guilefulness'
SHA256: 0000d79e1c6964e6806e9bbdaaaecb63dfabdb498f72bf28944119de1fe90d63
Binary prefix: 00000000000000001101011110011110...


### Explanation of Task 6: Proof of Work

This task demonstrates a core concept behind blockchain proof-of-work algorithms. In cryptocurrencies like Bitcoin. Bitcoin miners compete to find a nonce that, when combined with block data and hashed, produces a result with a specific number of leading zeros.

#### Concept:
- Finding inputs that produce hash outputs with specific patterns is computationally intensive
- There's no known way to predict which inputs will produce hashes with leading zeros
- The only practical approach is to try many different inputs (brute force)
- The probability of finding n leading zeros is approximately 1 in 2^n

#### Implementation:
- Instead of adding a nonce, we're searching through existing English words
- We calculate SHA256 hashes for a large sample of words
- We track words whose hash digests have the most leading zero bits
- This simulates a simplified version of blockchain mining

This demonstrates why proof-of-work is effective for consensus mechanisms - finding such patterns requires significant computational effort, but verifying them is trivial.

## Task 7: Turing Machines

### What Is a Turing Machine?

A Turing machine is a simple, abstract computational model introduced by Alan Turing in 1936. It consists of:

1. **An infinite tape** divided into discrete cells, each of which can hold a symbol from a finite alphabet (e.g., `0`, `1`, and a blank symbol `_`).
2. **A tape head** that can read the symbol in the current cell, write a new symbol in that cell, and move one cell left or right.
3. **A finite set of states**, including a designated start state and one or more halting (accept/reject) states.
4. **A transition function** that, based on the current state and the symbol under the head, specifies:
   - Which symbol to write,
   - Which direction to move the head (left or right),
   - Which state to transition into next.

Despite its simplicity, the Turing machine is powerful enough to model any computation that can be performed by modern computers. It serves as the foundation for the theory of computation and the concept of algorithmic decidability.


### Source

- https://plato.stanford.edu/entries/turing-machine/
- https://www.geeksforgeeks.org/turing-machine-in-toc/

In [95]:
def turingMachine(tape):
    # Convert the tape to a list for mutability
    tape = list(tape)
    
    # Find the starting position (leftmost non-blank symbol)
    head = 0
    while head < len(tape) and tape[head] == '_':
        head += 1
    
    # If the tape is all blanks, return "1"
    if head >= len(tape):
        return "1"
    
    state = "q0"  # Initial state
    
    while state != "q2":
        # Current symbol under the head
        current_symbol = tape[head] if 0 <= head < len(tape) else '_'
        
        if state == "q0":  # Moving right to find the end
            if current_symbol in ("0", "1"):
                head += 1
            elif current_symbol == "_":  # Found the end
                head -= 1
                state = "q1"
        
        elif state == "q1":  # Processing the addition
            if current_symbol == "0":
                # If we find a 0, change it to 1 and we're done (no carry)
                if 0 <= head < len(tape):
                    tape[head] = "1"
                state = "q2"  # Halt
            
            elif current_symbol == "1":
                # If we find a 1, change it to 0 and continue left (carry)
                if 0 <= head < len(tape):
                    tape[head] = "0"
                head -= 1
            
            elif current_symbol == "_":
                # We've moved past the beginning, need to add a new digit
                # This happens when the original number was all 1's
                if head < 0:
                    # Insert a 1 at the beginning
                    tape.insert(0, "1")
                    head = 0
                else:
                    tape[head] = "1"
                state = "q2"  # Halt
    
    # Remove leading blanks if any
    result = "".join(tape)
    result = result.lstrip('_')
    
    return result

# Example usage of the Turing machine
input_tape = "100111_"
result = turingMachine(input_tape)
print(f"Input: {input_tape.rstrip('_')}")
print(f"Output: {result.rstrip('_')}")

Input: 100111
Output: 101000


### Explanation of Turing task

This task implements a Turing Machine that adds 1 to a binary number. A Turing Machine is a mathematical model of computation consisting of:

- A tape divided into cells (our binary number plus a blank symbol '_')
- A head that can read and write symbols on the tape and move left or right
- A state register that stores the state of the Turing machine
- A finite table of instructions

States:

q0: Initial state - scan right to find the end of the number

q1: Add-one state - working from right to left, adding 1 with carry as needed

q2: Final state - computation complete

Algorithm:

1. Start at the leftmost digit in state q0  
2. Move right until reaching the blank symbol '_' (end of number)  
3. Move left one position to the rightmost digit (least significant bit)  
4. Switch to state q1 and begin addition:  
    - If current bit is 0: change to 1 and terminate (no carry)  
    - If current bit is 1: change to 0 and move left (carry the 1)  
5. If we reach the blank symbol at the left: add a new 1 (as the most significant bit)  
6. When addition is complete, switch to state q2 and terminate  
7. The implementation handles edge cases like carrying beyond the most significant bit by adding a new digit when necessary.  


## Task 8: Computational Complexity

In [96]:
import itertools

def bubble_sort(arr):
    # Create a copy to avoid modifying the original list
    arr = arr.copy()
    n = len(arr)
    comparisons = 0
    
    for i in range(n):
        # Flag to optimize if no swaps occur in a pass
        swapped = False
        
        for j in range(0, n - i - 1):
            comparisons += 1  # Count comparison
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
                
        # If no swapping occurred in this pass, array is sorted
        if not swapped:
            break
            
    return arr, comparisons

# Original list
L = [1, 2, 3, 4, 5]

# Generate all permutations
all_permutations = list(itertools.permutations(L))

# Sort each permutation and count comparisons
print(f"Total number of permutations: {len(all_permutations)}")
print("Permutation -> Number of comparisons")
print("-" * 40)

for perm in all_permutations:
    perm = list(perm)
    _, count = bubble_sort(perm)
    print(f"{perm} -> {count}")

# Let's also calculate some statistics
counts = [bubble_sort(list(perm))[1] for perm in all_permutations]
print("-" * 40)
print(f"Average comparisons: {sum(counts)/len(counts):.2f}")
print(f"Minimum comparisons: {min(counts)}")
print(f"Maximum comparisons: {max(counts)}")

Total number of permutations: 120
Permutation -> Number of comparisons
----------------------------------------
[1, 2, 3, 4, 5] -> 4
[1, 2, 3, 5, 4] -> 7
[1, 2, 4, 3, 5] -> 7
[1, 2, 4, 5, 3] -> 9
[1, 2, 5, 3, 4] -> 7
[1, 2, 5, 4, 3] -> 9
[1, 3, 2, 4, 5] -> 7
[1, 3, 2, 5, 4] -> 7
[1, 3, 4, 2, 5] -> 9
[1, 3, 4, 5, 2] -> 10
[1, 3, 5, 2, 4] -> 9
[1, 3, 5, 4, 2] -> 10
[1, 4, 2, 3, 5] -> 7
[1, 4, 2, 5, 3] -> 9
[1, 4, 3, 2, 5] -> 9
[1, 4, 3, 5, 2] -> 10
[1, 4, 5, 2, 3] -> 9
[1, 4, 5, 3, 2] -> 10
[1, 5, 2, 3, 4] -> 7
[1, 5, 2, 4, 3] -> 9
[1, 5, 3, 2, 4] -> 9
[1, 5, 3, 4, 2] -> 10
[1, 5, 4, 2, 3] -> 9
[1, 5, 4, 3, 2] -> 10
[2, 1, 3, 4, 5] -> 7
[2, 1, 3, 5, 4] -> 7
[2, 1, 4, 3, 5] -> 7
[2, 1, 4, 5, 3] -> 9
[2, 1, 5, 3, 4] -> 7
[2, 1, 5, 4, 3] -> 9
[2, 3, 1, 4, 5] -> 9
[2, 3, 1, 5, 4] -> 9
[2, 3, 4, 1, 5] -> 10
[2, 3, 4, 5, 1] -> 10
[2, 3, 5, 1, 4] -> 10
[2, 3, 5, 4, 1] -> 10
[2, 4, 1, 3, 5] -> 9
[2, 4, 1, 5, 3] -> 9
[2, 4, 3, 1, 5] -> 10
[2, 4, 3, 5, 1] -> 10
[2, 4, 5, 1, 3] -> 10
[2, 4, 5, 3, 1

### Explanation of Task 8

This task analyzes the computational complexity of bubble sort by counting comparisons across different input permutations.

#### Bubble Sort Algorithm:

Repeatedly steps through the list, comparing adjacent elements and swapping them if they're in the wrong order

With each pass, the largest unsorted element "bubbles" to its correct position
Continues until no swaps are needed in a pass, indicating the list is sorted

#### Implementation Details:

The optimized bubble sort includes an early termination flag (swapped)
A counter tracks each comparison operation

All 120 permutations (5!) of [1,2,3,4,5] are generated and sorted
For each permutation, we record the number of comparisons required

#### Analysis of Results:

- Best case: 4 comparisons (for already sorted list [1,2,3,4,5])

- Worst case: 10 comparisons (for lists with elements far from their sorted positions)

- Average case: 9.26 comparisons

This demonstrates how the initial ordering significantly affects bubble sort's performance:

Already sorted or nearly sorted lists perform much better

Reverse-sorted lists or those with elements far from their final positions perform worse

Bubble sort has O(n²) worst-case time complexity, but can achieve O(n) in the best case

The variation in comparison counts clearly illustrates why bubble sort is inefficient for large datasets but may be suitable for small or nearly-sorted lists.